modules/storages/app/services/storages/file_links/copy_file_links_service.rb

```diff
@@ -0,0 +1,111 @@
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  module FileLinks
+    class CopyFileLinksService
+      def self.call(source:, target:, user:, work_packages_map:)
+        new(source:, target:, user:, work_packages_map:).call
+      end
+
+      def initialize(source:, target:, user:, work_packages_map:)
+        @source = source
+        @target = target
+        @user = user
+        @work_packages_map = work_packages_map.to_h { |key, value| [key.to_i, value.to_i] }
+      end
+
+      def call
+        source_file_links = FileLink
+          .includes(:creator)
+          .where(container_id: @work_packages_map.keys, container_type: "WorkPackage")
+
+        return create_unmanaged_file_links(source_file_links) if @source.project_folder_manual?
+
+        create_managed_file_links(source_file_links)
+      end
+
+      private
+
+      def create_managed_file_links(source_file_links)
+        target_files = target_files_map.result
+        source_files = source_files_info(source_file_links).result
+
+        location_map = build_location_map(source_files, target_files)
+
+        source_file_links.find_each do |source_link|
+          attributes = source_link.dup.attributes
+
+          attributes.merge!(
+            'creator_id' => @user.id,
+            'container_id' => @work_packages_map[source_link.container_id],
+            'origin_id' => location_map[source_link.origin_id]
+          )
+
+          FileLinks::CreateService.new(user: @user).call(attributes)
+        end
+      end
+
+      def build_location_map(source_files, target_location_map)
+        source_location_map = source_files.to_h { |info| [info.id, info.location] }
+
+        source_location_map.each_with_object({}) do |(id, location), output|
+          target = location.gsub(@source.managed_project_folder_path, @target.managed_project_folder_path)
+
+          output[id] = target_location_map[target]
+        end
+      end
+
+      def source_files_info(source_file_links)
+        Peripherals::Registry
+          .resolve("#{@source.storage.short_provider_type}.queries.files_info")
+          .call(storage: @source.storage, user: @user, file_ids: source_file_links.pluck(:origin_id))
+      end
+
+      def target_files_map
+        Peripherals::Registry
+          .resolve("#{@source.storage.short_provider_type}.queries.folder_files_file_ids_deep_query")
+          .call(storage: @source.storage, folder: Peripherals::ParentFolder.new(@target.project_folder_location))
+      end
+
+      def create_unmanaged_file_links(source_file_links)
+        source_file_links.find_each do |source_file_link|
+          attributes = source_file_link.dup.attributes
+
+          attributes['creator_id'] = @user.id
+          attributes['container_id'] = @work_packages_map[source_file_link.container_id]
+
+          # TODO: Do something when this fails
```

Comment:
🟡 resolve todo? 😁 

This is not part of current implementation, I know. What should we do?

`CopyFileLinksService` seems to not return a `ServiceResult` right now. Should it return one and log the included failures from the caller?

Or simply log the failures when the creation fails.

Reply:
That was part of the feedback I wanted. I think for now I'll just log the errors.

We need to decide on what to return when some object does multiple operations, some of which might fail but not be catastrophic. A success with errors feels about right, but I'm no sure. Same goes of for having a result on a failure, feels correct but... idk.